pkg/storage/unified/resource/storage_backend_test.go

```diff
@@ -0,0 +1,1030 @@
+	originalObj, err := createTestObjectWithName("test-resource", "apps", "initial-data")
+	require.NoError(t, err)
+	require.Equal(t, objectToJSONBytes(t, originalObj), collectedItems[0])
+}
+
+func TestKvStorageBackend_ListHistory_Success(t *testing.T) {
+	backend := setupTestStorageBackend(t)
+	ctx := context.Background()
+
+	// Create initial resource
+	testObj, err := createTestObjectWithName("test-resource", "apps", "initial-data")
+	require.NoError(t, err)
+
+	metaAccessor, err := utils.MetaAccessor(testObj)
+	require.NoError(t, err)
+
+	writeEvent := WriteEvent{
+		Type: resourcepb.WatchEvent_ADDED,
+		Key: &resourcepb.ResourceKey{
+			Namespace: "default",
+			Group:     "apps",
+			Resource:  "deployments",
+			Name:      "test-resource",
+		},
+		Value:      objectToJSONBytes(t, testObj),
+		Object:     metaAccessor,
+		PreviousRV: 0,
+	}
+
+	rv1, err := backend.WriteEvent(ctx, writeEvent)
+	require.NoError(t, err)
+
+	// Update the resource
+	testObj.Object["spec"].(map[string]any)["value"] = "updated-data"
+	writeEvent.Type = resourcepb.WatchEvent_MODIFIED
+	writeEvent.Value = objectToJSONBytes(t, testObj)
+	writeEvent.PreviousRV = rv1
+
+	rv2, err := backend.WriteEvent(ctx, writeEvent)
+	require.NoError(t, err)
+
+	// Update again
+	testObj.Object["spec"].(map[string]any)["value"] = "final-data"
+	writeEvent.Value = objectToJSONBytes(t, testObj)
+	writeEvent.PreviousRV = rv2
+
+	rv3, err := backend.WriteEvent(ctx, writeEvent)
+	require.NoError(t, err)
+
+	// List the history
+	listReq := &resourcepb.ListRequest{
+		Options: &resourcepb.ListOptions{
+			Key: &resourcepb.ResourceKey{
+				Namespace: "default",
+				Group:     "apps",
+				Resource:  "deployments",
+				Name:      "test-resource",
+			},
+		},
+		Source: resourcepb.ListRequest_HISTORY,
+		Limit:  10,
+	}
+
+	var historyItems []struct {
+		resourceVersion int64
+		value           []byte
+	}
+
+	rv, err := backend.ListHistory(ctx, listReq, func(iter ListIterator) error {
+		for iter.Next() {
+			if err := iter.Error(); err != nil {
+				return err
+			}
+			historyItems = append(historyItems, struct {
+				resourceVersion int64
+				value           []byte
+			}{
+				resourceVersion: iter.ResourceVersion(),
+				value:           iter.Value(),
+			})
+		}
+		return iter.Error()
+	})
+
+	require.NoError(t, err)
+	require.Greater(t, rv, int64(0))
+	require.Len(t, historyItems, 3) // Should have all 3 versions
+
+	// Verify the history is sorted (newest first by default)
+	require.Equal(t, rv3, historyItems[0].resourceVersion)
+	require.Equal(t, rv2, historyItems[1].resourceVersion)
+	require.Equal(t, rv1, historyItems[2].resourceVersion)
+
+	// Verify the content matches expectations
+	finalObj, err := createTestObjectWithName("test-resource", "apps", "final-data")
+	require.NoError(t, err)
+	require.Equal(t, objectToJSONBytes(t, finalObj), historyItems[0].value)
+}
+
+func TestKvStorageBackend_ListTrash_Success(t *testing.T) {
```

Comment:
Is there a test for listing "trash" in case when there's no deleted version, or object is "alive" again?

Reply:
runTestIntegrationBackendTrash